src/lib.rs

```diff
@@ -1010,11 +1324,376 @@ decl_module! {
+
+        // No ancestor is blocking us doing mutation in this category
+        Self::ensure_can_mutate_in_path_leaf(&category_tree_path)?;
+
+        // Make and add initial post
+        let new_post_id = NextPostId::get();
+
+        let new_post = Post {
+            id: new_post_id,
+            thread_id: thread_id,
+            current_text: text.clone(),
+            moderation: None,
+            text_change_history: vec![],
+            created_at: Self::current_block_and_time(),
+            author_id: author_id,
+            nr_in_thread: thread.num_posts_ever_created(),
+        };
+
+        // Store post
+        <PostById<T>>::insert(new_post_id, new_post.clone());
+
+        // Update next post id
+        NextPostId::mutate(|n| {
+            *n += 1;
+        });
+
+        // Update unmoderated post count of thread
+        <ThreadById<T>>::mutate(thread_id, |t| {
+            t.num_unmoderated_posts += 1;
+        });
+
+        Ok(new_post)
+    }
+
+    // The method only called from other module to create a forum user.
+    fn create_forum_user(
+        account_id: T::AccountId,
+        name: Vec<u8>,
+        self_introduction: Vec<u8>,
+    ) -> dispatch::Result {
+        Self::ensure_user_name_is_valid(&name)?;
+
+        Self::ensure_user_self_introduction_is_valid(&self_introduction)?;
+
+        if <ForumUserIdByAccount<T>>::exists(account_id.clone()) {
+            return Err(ERROR_USER_ALREADY_REGISTERED_FORUM);
+        }
+
+        let new_forum_user = ForumUser {
+            id: account_id.clone(),
+            name: name.clone(),
+            self_introduction: self_introduction.clone(),
+        };
+
+        <ForumUserById<T>>::insert(NextForumUserId::get(), new_forum_user);
+        <ForumUserIdByAccount<T>>::insert(account_id, NextForumUserId::get());
+
+        NextForumUserId::mutate(|n| {
+            *n += 1;
+        });
+        Ok(())
+    }
+
+    // The method only called from other module to create a new moderator.
+    fn create_moderator(
+        account_id: T::AccountId,
+        name: Vec<u8>,
+        self_introduction: Vec<u8>,
+    ) -> dispatch::Result {
+        Self::ensure_user_name_is_valid(&name)?;
+
+        Self::ensure_user_self_introduction_is_valid(&self_introduction)?;
+
+        if <ModeratorIdByAccount<T>>::exists(account_id.clone()) {
+            return Err(ERROR_USER_ALREADY_REGISTERED_MODERATOR);
+        }
+
+        let new_moderator = Moderator {
+            id: account_id.clone(),
+            name: name,
+            self_introduction: self_introduction,
+        };
+
+        <ModeratorById<T>>::insert(NextModeratorId::get(), new_moderator);
+        <ModeratorIdByAccount<T>>::insert(account_id, NextModeratorId::get());
+
+        NextModeratorId::mutate(|n| {
+            *n += 1;
+        });
+        Ok(())
+    }
+
+    // The method only called from other module to add some labels.
+    fn add_labels(labels: Vec<Vec<u8>>) -> dispatch::Result {
```

Comment:
According to the requirement, the new labels should be introduced through a non-dispatchable call. So I implement it in this way. I will put the moderator add new labels as new function, and commit it with new PR in the future since the PR already included too much code change.

https://github.com/Joystream/substrate-forum-module/issues/45
There is a set of available labels, and each thread and category can select some number of existing labels. New labels are introduced through a non-dispatchable call. The labels applied to a given thread or category can be edited, not only by the author, but also by the appropriate moderators.
Remember to introduce a constraint variable: max number of labels applied to a single thread or post.